airflow/configuration.py

```diff
@@ -393,31 +391,38 @@ def mkdir_p(path):
 else:
     AIRFLOW_CONFIG = expand_env_var(os.environ['AIRFLOW_CONFIG'])
 
-if not os.path.isfile(AIRFLOW_CONFIG):
+
+def default_config():
     """
-    These configuration options are used to generate a default configuration
-    when it is missing. The right way to change your configuration is to alter
-    your configuration file, not this code.
+        Generates a default configuration from the default template + currently
+         defined variables
     """
+
     FERNET_KEY = generate_fernet_key()
-    logging.info("Creating new config file in: " + AIRFLOW_CONFIG)
-    f = open(AIRFLOW_CONFIG, 'w')
-    f.write(DEFAULT_CONFIG.format(**locals()))
-    f.close()
+    vars = {k: v for d in [globals(), locals()] for k, v in d.items()}
+    return DEFAULT_CONFIG.format(**vars)
 
 TEST_CONFIG_FILE = AIRFLOW_HOME + '/unittests.cfg'
 if not os.path.isfile(TEST_CONFIG_FILE):
     logging.info("Creating new config file in: " + TEST_CONFIG_FILE)
-    f = open(TEST_CONFIG_FILE, 'w')
-    f.write(TEST_CONFIG.format(**locals()))
-    f.close()
+    with open(TEST_CONFIG_FILE, 'w') as f:
+        f.write(TEST_CONFIG.format(**locals()))
 
-logging.info("Reading the config from " + AIRFLOW_CONFIG)
+if not os.path.isfile(AIRFLOW_CONFIG):
+    """
+    These configuration options are used to generate a default configuration
+    when it is missing. The right way to change your configuration is to alter
+    your configuration file, not this code.
+    """
+    logging.info("Creating new config file in: " + AIRFLOW_CONFIG)
+    with open(AIRFLOW_CONFIG, 'w') as f:
+        f.write(default_config())
 
+logging.info("Reading the config from " + AIRFLOW_CONFIG)
 
 def test_mode():
     conf = ConfigParserWithDefaults(defaults)
-    conf.read(TEST_CONFIG)
+    conf.read(TEST_CONFIG_FILE)
```

Comment:
> (The github split view will help more then the unified view here)
> We’ve fixed several problems up here : 
> If `AIRFLOW_CONFIG = unittest.cfg`, we’re now first generating the unittest.cfg file from the `TEST_CONFIG` rather than using `DEFAULT_CONFIG`
> `test_mode()` now reads correctly from the file, it was failing silently before